journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+              findFreeObject(state) match {
+                case Some((id, obj)) =>
+                  for {
+                    _ <- markObjectAsUsed(id, obj)
+                    _ <- promise.complete((id, obj.value).asRight)
+                  } yield ()
+
+                case None =>
+                  if (state.objects.size < poolSize) {
+                    for {
+                      o <- allocateObject(state)
+                      _ <- promise.complete(o.asRight)
+                    } yield ()
+                  } else {
+                    registerWaiter(state, promise)
+                  }
+              }
+            }
+          }.as(promise)
+        }
+
+      val acquireObject: F[(ObjectId, O)] = {
+        acquirePromise.flatMap { promise =>
+          promise.get.flatMap {
+            case Right(x) => x.pure[F]
+            case Left(err) => Async[F].raiseError(err)
+          }
+        }
+      }
+
+      def checkIfExpired(id: ObjectId, timeoutScheduledAt: FiniteDuration): F[Unit] = {
+        mutex.lock.surround {
+          stateRef.get.map(_.objects.get(id)).flatMap {
+            case Some(AllocatedObject(_, deallocate, AllocatedObject.State.Free(releasedAt))) =>
+              if (releasedAt == timeoutScheduledAt) {
+                for {
+                  _ <- deallocate.handleErrorWith { err =>
+                    log.error(s"Failed to deallocate resource $id", err)
+                  }.start
+
+                  _ <- stateRef.update { state =>
+                    state.copy(objects = state.objects - id)
+                  }
+                } yield ()
+              } else {
+                Async[F].unit
+              }
+
+            case o =>
+              log.error(
+                s"Resource $id is in unexpected state ${o.map(_.state)}. This is not normal, please contact the maintainers."
+              )
+          }
+        }
+      }
+
+      def releaseObject(id: ObjectId): F[Unit] =
+        mutex.lock.surround {
+          stateRef.get.flatMap { state =>
+            state.objects.get(id) match {
+              case None =>
+                log.error(s"Resource $id was returned to the pool but it was not found in the state. " +
+                  "This is not normal, please contact the maintainers.")
+
+              case Some(obj) =>
+                state.waiters.headOption match {
+                  case Some((waiterId, waiter)) =>
+                    for {
+                      _ <- waiter.complete((id, obj.value).asRight[Throwable])
+                      _ <- stateRef.update { state =>
+                        state.copy(waiters = state.waiters - waiterId)
+                      }
+                      _ <- markObjectAsUsed(id, obj)
+                    } yield ()
+
+                  case None =>
+                    for {
+                      releasedAt <- Async[F].monotonic
+                      _ <- (Async[F].sleep(resourceTTL) *> checkIfExpired(id, releasedAt)).start
```

Comment:
This will produce a memory leak if one and the same resource is acquired / released a lot. I.e. each release will use a bit of memory, which will only be released when `resourceTTL` is reached. I am not sure it is important, but still decided to mention it.

Reply:
I see, good point. What do you propose? Periodically check for expired resources across the entire pool instead of scheduling a check every time?

Reply:
I don't think this can be a bottleneck. How many releases of the same resource should happen for this code to cause an OOM? 50,000? I tested locally with ResourcePoolSpec by changing `timesToAcquire` parameter. My pool did fine with 50,000 acquires. Even for 500,000 acquires there was no OOM, it was just slow.

Reply:
I'd agree with @rtar on this one. 
I'd also prefer periodic background job that will expire un-used resources.


Imho that's too much of overhead to schedule timer per `access`